src/System.Runtime.Extensions/src/System/IO/PathHelper.Windows.cs

```diff
@@ -2,457 +2,373 @@
+        private const int MAX_BUILDER_SIZE = 1024;
 
-        // Instantiates a PathHelper with a heap alloc'd array of ints. Will create a StringBuilder
-        internal PathHelper(int capacity, int maxPath)
-        {
-            this.m_sb = new StringBuilder(capacity);
-            this.m_capacity = capacity;
-            this.m_maxPath = maxPath;
-            doNotTryExpandShortFileName = false;
-        }
+        // Can't be over 8.3 and be a short name
+        private const int MAX_SHORT_NAME = 12;
 
-        internal int Length
-        {
-            get
-            {
-                if (useStackAlloc)
-                {
-                    return m_length;
-                }
-                else
-                {
-                    return m_sb.Length;
-                }
-            }
-            set
-            {
-                if (useStackAlloc)
-                {
-                    m_length = value;
-                }
-                else
-                {
-                    m_sb.Length = value;
-                }
-            }
-        }
+        private const char LastAnsi = (char)255;
+        private const char Delete = (char)127;
 
-        internal int Capacity
-        {
-            get
-            {
-                return m_capacity;
-            }
-        }
+        // Trim trailing white spaces, tabs etc but don't be aggressive in removing everything that has UnicodeCategory of trailing space.
+        // string.WhitespaceChars will trim more aggressively than what the underlying FS does (for ex, NTFS, FAT).
+        private static readonly char[] TrimEndChars = { (char)0x9, (char)0xA, (char)0xB, (char)0xC, (char)0xD, (char)0x20, (char)0x85, (char)0xA0 };
 
-        internal char this[int index]
+        // We don't use StringBuilderCache because the cached size is too small and we don't want to collide with other usages.
+        [ThreadStatic]
+        private static StringBuilder t_cachedOutputBuffer;
+
+        internal string Normalize(string path, bool fullCheck, bool expandShortPaths)
         {
-            [System.Security.SecurityCritical]
-            get
+            try
             {
-                Contract.Requires(index >= 0 && index < Length);
-                if (useStackAlloc)
-                {
-                    return m_arrayPtr[index];
-                }
-                else
-                {
-                    return m_sb[index];
-                }
+                return this.NormalizeInternal(path, fullCheck, expandShortPaths);
             }
-            [System.Security.SecurityCritical]
-            set
+            finally
             {
-                Contract.Requires(index >= 0 && index < Length);
-                if (useStackAlloc)
-                {
-                    m_arrayPtr[index] = value;
-                }
-                else
+                if (t_cachedOutputBuffer != null && t_cachedOutputBuffer.Capacity > MAX_BUILDER_SIZE)
```

Comment:
It is all contained here so it shouldn't be possible to hit multiple accesses, right? I was trying to avoid an extra nest on the main loop for clarity. I'll flip things around.